packages/otel/src/instrumentations/fetch.ts

```diff
@@ -412,12 +640,92 @@ function onError(span: Span, err: unknown): void {
 function headersToAttributes(
   span: Span,
   attrsToHeadersMap: Record<string, string>,
-  headers: Headers
+  headers: Headers | IncomingHttpHeaders | OutgoingHttpHeaders,
+  getter: TextMapGetter<Headers | IncomingHttpHeaders | OutgoingHttpHeaders>
 ): void {
   for (const [attrName, headerName] of Object.entries(attrsToHeadersMap)) {
-    const headerValue = headers.get(headerName);
-    if (headerValue !== null) {
+    const headerValue = getter.get(headers, headerName);
+    if (headerValue !== undefined) {
       span.setAttribute(attrName, headerValue);
     }
   }
 }
+
+function constructUrlFromRequestOptions(options: http.RequestOptions, protocolFromModule: string): URL {
+  if (options.socketPath) {
+    throw new Error(
+      'Cannot construct a network URL: options.socketPath is specified, indicating a Unix domain socket.'
+    );
+  }
+
+  let protocol = options.protocol ?? protocolFromModule;
+
+  if (protocol && !protocol.endsWith(':')) {
+    protocol += ':';
+  }
+
+  // Per documentation: "hostname will be used if both host and hostname are specified."
+  // If options.hostname is present, it takes precedence.
+  // If options.hostname is absent, options.host is used (which might contain 'hostname:port').
+  let hostname = options.hostname;
+  let port = options.port ?? options.defaultPort; // Can be string (e.g. from URL.port), number, or undefined
+
+  if (!hostname && options.host) {
+    // Try to parse hostname and port from options.host
+    const hostParts = options.host.split(':');
+    hostname = hostParts[0];
+    const portPart = hostParts[1];
+    if (hostParts.length > 1 && portPart && port === undefined) {
+      const parsedPort = parseInt(portPart, 10);
+      if (!isNaN(parsedPort)) {
+        port = parsedPort;
+      }
+    }
+  }
+
+  // If hostname is still not determined (e.g. options.host was also undefined or only a port like ':8080')
+  // use default 'localhost' as per http.request behavior for options.host.
+  if (!hostname) {
+    hostname = 'localhost';
+  }
+
+  if (!hostname) { // Should not happen if default to localhost is applied
```

Comment:
Can't happen? It's right above.

Reply:
good spot! removed here: https://github.com/vercel/otel/pull/159/commits/b66e529e1a71ab5ced9425e595de369efbcf895a